pkg/test/client.go

```diff
@@ -87,6 +98,27 @@ func (c *FakeClient) Update(ctx context.Context, obj runtime.Object, opts ...cli
 	if c.MockUpdate != nil {
 		return c.MockUpdate(ctx, obj, opts...)
 	}
+
+	// Update Generation if needed since the kube fake client doesn't update generations.
+	// Compare the specs (only) and only increment the generation if something changed
+	// (the server will check the object metadata, but we're skipping this here)
+	if svc, ok := obj.(*corev1.Service); ok {
+		existing := corev1.Service{}
+		if err := c.Client.Get(ctx, types.NamespacedName{Namespace: svc.GetNamespace(), Name: svc.GetName()}, &existing); err != nil {
+			return err
+		}
+		if !reflect.DeepEqual(existing.Spec, svc.Spec) { // Service has a `spec` field
+			svc.SetGeneration(existing.GetGeneration() + 1)
+		}
+	} else if cm, ok := obj.(*corev1.ConfigMap); ok {
+		existing := corev1.ConfigMap{}
+		if err := c.Client.Get(ctx, types.NamespacedName{Namespace: cm.GetNamespace(), Name: cm.GetName()}, &existing); err != nil {
+			return err
+		}
+		if !reflect.DeepEqual(existing.Data, cm.Data) { // ConfigMap has a `data` field
+			cm.SetGeneration(existing.GetGeneration() + 1)
+		}
+	}
```

Comment:
This should work and should be more generic:
```go
updatingMeta, err := meta.Accessor(obj)
if err != nil {
	return err
}

current := obj.DeepCopyObject()
if err := c.Client.Get(ctx, types.NamespacedName{Namespace: updatingMeta.GetNamespace(), Name: updatingMeta.GetName()}, current); err != nil {
	return err
}
currentMeta, err := meta.Accessor(current)
if err != nil {
	return err
}

updatingContent, err := json.Marshal(obj)
if err != nil {
	return err
}
updatingMap := map[string]interface{}{}
if err := json.Unmarshal(updatingContent, &updatingMap); err != nil {
	return err
}
currentContent, err := json.Marshal(current)
if err != nil {
	return err
}
currentMap := map[string]interface{}{}
if err := json.Unmarshal(currentContent, &currentMap); err != nil {
	return err
}

if updatingMap["spec"] != nil {
	if !reflect.DeepEqual(updatingMap["spec"], currentMap["spec"]) {
		updatingMeta.SetGeneration(currentMeta.GetGeneration() + 1)
	}
} else if updatingMap["data"] != nil {
	if !reflect.DeepEqual(updatingMap["data"], currentMap["data"]) {
		updatingMeta.SetGeneration(currentMeta.GetGeneration() + 1)
	}
}
```

Reply:
Thanks! It doesn't cover all the use cases but works most of them.
Updated in https://github.com/codeready-toolchain/toolchain-common/pull/89/commits/e2334fbfe888e9c8daddf1610b4bc730e3240e21

Reply:
we could maybe cover more cases with this version
```go
	updatingMap["metadata"] = nil
	updatingMap["status"] = nil
	updatingMap["kind"] = nil
	updatingMap["apiVersion"] = nil

	currentMap["metadata"] = nil
	currentMap["status"] = nil
	currentMap["kind"] = nil
	currentMap["apiVersion"] = nil

	if !reflect.DeepEqual(updatingMap, currentMap) {
		updatingMeta.SetGeneration(currentMeta.GetGeneration() + 1)
	}
```

Reply:
Yes, it's a good idea! I also added some cleanup for the object we use an empty one for loading the current resource to avoid the situation when we have some leftovers: https://github.com/codeready-toolchain/toolchain-common/pull/89/commits/c4811ce8458f9c9cfabc627dafe31c46fd1515f3